infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/NamespaceNameValidator.java

```diff
@@ -51,6 +53,18 @@ public static boolean isValid(String name) {
     return validateInternal(name).isOk();
   }
 
+  public static String normalize(String namespaceName) {
```

Comment:
As this seems to be a good stretching exercise in the morning, here's my impl :D
```
  public static String normalize(String namespaceName) {
    namespaceName = namespaceName.replaceAll("[^-a-zA-Z0-9]", "-"); // replace invalid chars with '-'
    namespaceName = namespaceName.replaceAll("-+", "-");  // replace multiple '-' with single ones
    namespaceName = Strings.trimLeadingCharacter(namespaceName, '-'); // remove all '-' in the beginning
    namespaceName = Strings.trimTrailingCharacter(namespaceName, '-'); // remove all '-' in the end
    return namespaceName.substring(0, Math.min(namespaceName.length(), METADATA_NAME_MAX_LENGTH)); // limit length to METADATA_NAME_MAX_LENGTH
  }
```

Reply:
Mine would run circles around yours performance wise but since this is never gonna be run 50000 times per second, a :+1: from me as well :)

Reply:
I don't think so, this looks like infinite loop for me once it has valid character in the beginning :)
```
while (!sb.isEmpty()) {
  if (!isValidNamespaceNameStartOrEnd(sb.charAt(0))) {
    sb.remove(0);
  }
}
```

I agree that we could do something more performant, but I would choose readability here. Considering the fact, that we could fit like million iterations of this in one fetch namespace request to k8s API...